### PR TITLE
Hexagons and hexikes

### PR DESCRIPTION
**Pull request by Jordan-Dennis**
Hi all,
I have not allocated a position to the new `layers.py` which contains __only__ the new `HexagonalBasis` code and friends. This code has been visually tested very thoroughly against `poppy` although I have held of on writing formal unit tests on it because I _know_ it works and there are more pressing things to develop. 

The code is `jit`-able but due to the overhead incurred by the additional vectorizations I have made it is actually slower than the `numpy` version even after `jit`. This should be fine because we are just saving to a file with the `save(filename : str)` method. 

I have not introduced a `CompoundAperture` class to deal with saving multiple basis in tensors to files. RSVP me if we want to add this functionality. Currently we can rotate, shear (spelled correctly :laughing:), translate and magnify the basis vectors. 

As a final note; conceding to `jax` we have our first constant `MAX_DIFF`, if you want to generate more the 21 basis vectors you will have to increase this to 6 or above, but I doubt we will. For lower memory use and slightly better speed you can drop it to 2 if you are only generating the first nine. 

@LouisDesdoigts, if you can review the code after you have settled in :canada:, and leave a note allocating the position that you want me to put the code and I will move it there; I was thinking maybe a `basis.py` in the `dLux/dLux` directory.  

Regards 

Jordan

### CODE REVIEW

File: dLux/base.py [hunk 1]
```diff
@@ -91,7 +91,7 @@ def debug_prop(self, wavel, offset=np.zeros(2)):
         I believe this is diffable but there is no reason to force it to be
         """
         # params_dict = {"Wavefront": Wavefront(wavel, offset)}
-        params_dict = {"Wavefront": PhysicalWavefront(wavel, offset)}
+        params_dict = {"Wavefront": dLux.PhysicalWavefront(wavel, offset)}
```
Comment (LouisDesdoigts) [inline dLux/base.py:94]: I believe this change also needs to applied to `OpticalSystem.proagate_mono()`

File: dLux/utils/hexike.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    Returns
+    -------
+    x : Array
+        The pixel positions along the edge of the image.
+    """
+    return np.arange(number_of_pixels) - number_of_pixels / 2. \
+        + 0.5 + pixel_offset
+    
+
+def _get_pixel_positions(
```
Comment (LouisDesdoigts) [inline dLux/utils/hexike.py:36]: This function should be imported from coordinates.py

File: dLux/utils/zernike.py [hunk 3]
```diff
@@ -217,7 +217,7 @@ def R(n, m, rho):
     n, m : int
         Zernike function degree
     rho : array
-        Image plane radial coordinates. `rho` should be 1 at the desired pixel radius of the
+        Image plane radial coordnates. `rho` should be 1 at the desired pixel radius of the
```
Comment (LouisDesdoigts) [inline dLux/utils/zernike.py:220]: small typo here: 'coordnates' in stead of 'coordinates'

**Comment by LouisDesdoigts** [comment]
High level comments:

dLux/coordinates.py
> I would add `get_pixel_coordinates` also, which simply multiplies the result of `get_pixel_positions`
>
> Similarly it would make sense to use the functions in utils.py script to deal with the cartesian -> polar transform, and probably roll the functionality of `get_radial_positions` into `get_pixel_coordinates` with some `polar=False` input or something like that - doesn't need to be a new function.

**Comment by Jordan-Dennis** [comment]
The `dLux/utils/hexike.py` is the old code from the early development. The new code is under `dLux/dev/layers.py` (name needs to be changed). This is the code that can do the rotations ect. I might actually delete the functional version, but I think it is worth keeping the `zernike.py` file because it is faster if that is all you want to use.

**Pull request merged**